packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -309,267 +333,176 @@ class FadeInImage extends StatefulWidget {
-    return _isShowingPlaceholder
-      ? widget.placeholderSemanticLabel
-      : widget.imageSemanticLabel;
   }
 
-  @override
-  Widget build(BuildContext context) {
-    assert(_phase != FadeInImagePhase.start);
-    final ImageInfo imageInfo = _imageInfo;
-    final RawImage image = RawImage(
-      image: imageInfo?.image,
+  Image _image({
+    @required ImageProvider image,
+    @required ImageFrameBuilder frameBuilder,
+  }) {
+    assert(image != null);
+    assert(frameBuilder != null);
+    return Image(
+      image: image,
+      frameBuilder: frameBuilder,
       width: widget.width,
       height: widget.height,
-      scale: imageInfo?.scale ?? 1.0,
-      color: Color.fromRGBO(255, 255, 255, _animation?.value ?? 1.0),
-      colorBlendMode: BlendMode.modulate,
       fit: widget.fit,
       alignment: widget.alignment,
       repeat: widget.repeat,
       matchTextDirection: widget.matchTextDirection,
-    );
-
-    if (widget.excludeFromSemantics) {
-      return image;
-    }
-
-    return Semantics(
-      container: _semanticLabel != null,
-      image: true,
-      label: _semanticLabel ?? '',
-      child: image,
+      excludeFromSemantics: true,
     );
   }
 
   @override
-  void debugFillProperties(DiagnosticPropertiesBuilder description) {
-    super.debugFillProperties(description);
-    description.add(EnumProperty<FadeInImagePhase>('phase', _phase));
-    description.add(DiagnosticsProperty<ImageInfo>('pixels', _imageInfo));
-    description.add(DiagnosticsProperty<ImageStream>('image stream', _imageResolver._imageStream));
-    description.add(DiagnosticsProperty<ImageStream>('placeholder stream', _placeholderResolver._imageStream));
+  Widget build(BuildContext context) {
+    Widget result = Stack(
+      fit: StackFit.passthrough,
+      alignment: AlignmentDirectional.center,
+      // Text direction is irrelevant here since we're using center alignment,
+      // but it allows the Stack to avoid a call to Directionality.of()
+      textDirection: TextDirection.ltr,
+      children: <Widget>[
+        _image(
+          image: widget.image,
+          frameBuilder: (BuildContext context, Widget child, int frame, bool wasSynchronouslyLoaded) {
+            if (wasSynchronouslyLoaded && widget.skipFadeOnSynchronousLoad)
+              _controller.value = _controller.upperBound;
+            else if (frame == 0 && !_controller.isAnimating)
+              _controller.animateTo(1.0, duration: widget.fadeInDuration + widget.fadeOutDuration);
+            return FadeTransition(
+              opacity: _fadeInAnimation,
+              child: child,
+            );
+          },
+        ),
+        _image(
+          image: widget.placeholder,
+          frameBuilder: (BuildContext context, Widget child, int frame, bool wasSynchronouslyLoaded) {
+            return FadeTransition(
+              opacity: _fadeOutAnimation,
+              child: child,
+            );
+          },
+        ),
+      ],
+    );
```

Comment:
could we make FadeInImage do a proper cross-fade with BlendMode.plus the way Porter and Duff intended?

Reply:
Yep.  But it'd change the way this widget's fade behaves.  Worth it?

Reply:
I left the behavior as-is for this class.  I may follow-up with a separate `CrossFadeImage` if we think it's worth the maintenance costs.